### ISSUE TITLE
Documentation for map_batches / map_elements / map_groups is confusing 

### ISSUE THREAD

**Issue by soerenwolfers** [pola-rs/polars#14521]
## Summary

The functionality split between `map_batches` / `map_elements` / `map_groups` is complicated enough that it warrants having completely precise and rich-of-examples documentation. 

A great and very closely related example in this area is numpy's documentation of (generalized) ufuncs at https://numpy.org/doc/stable/reference/c-api/generalized-ufuncs.html

(Personally, I feel the API behind the polars functions itself might use some improvement as well, but that's another ticket, and not one that I as a novice should make claims about.)

Below, I list some specific gripes with the current documentation of the three functions, but I think beyond fixing all three, it'd be useful to have some joint documentation of the differences and use-cases of each of the various UDF functions:

### map_elements
[Documentation for map_elements](https://docs.pola.rs/py-polars/html/reference/expressions/api/polars.Expr.map_elements.html)

Confusingly, `map_elements` has a name that very clearly suggests that the function here is applied pointwise, or at least should operate in a pointwise fashion, yet the documentation indicates that it can be used to emulate grouped folds, such as `group_by().cum_sum()`, given that 
according to the documentation:
```
GroupBy
Expects function to be of type Callable[[Series], Any]. For each group, applies a Python function to the slice of the column corresponding to that group.
```
I think it should be more clearly highlighted that `map_elements` is allowed in a non-pointwise fashion in group-by contexts (or that it is not, in case this is an implementation detail that may change)

### map_groups

[Documentation for map_groups](https://docs.pola.rs/py-polars/html/reference/expressions/api/polars.map_groups.html)
- The description says nothing but "Apply a custom/user-defined function (UDF) in a GroupBy context." which leaves the question why this function exists, given that `map_elements` can also apply a UDF in a GroupBy context
- The documentation says nothing about the `"function"` parameter except that it contradicts what's in the signature, i.e., that it is fed a Sequence of Series.

### map_batches
[Documentation for map_batches](https://docs.pola.rs/py-polars/html/reference/expressions/api/polars.Expr.map_batches.html)  

This is the worst. 

- First, nowhere in the documentation is it even hinted at what a "batch" is. Instead, the documentation starts by saying
```
Apply a custom python function to a whole Series or sequence of Series.
```
yet, the author says at https://github.com/pola-rs/polars/issues/12941 that
```
A batch by definition doesn't include a whole group. E.g. if you take a batch from a data set and you sort that batch, you shouldn't assume the whole dataset is sorted.
```
Being new to polars, I don't understand what this means (that's not a complaint, issues don't have to use language for beginners), and confusingly [in another place](https://github.com/pola-rs/polars/issues/10744 ) he seems to contradict (but this may just be my misunderstanding) even that, saying that `map_batches` doesn't even get fed complete "batches":
```
> Methods are applied to the object they are called on. So if I call map on a Series, I expect to map the whole Series.

This is not guaranteed. Expr.map may get an incomplete batch. This happens for instance in the streaming engine.
```

- Next, the documentation says 
```
Warning

If you are looking to map a function over a window function or group_by context, refer to `map_elements()` instead.
```
yet the description of the `agg_list` parameter says
```
Aggregate the values of the expression into a list before applying the function. This parameter only works in a group-by context
```
and half the documentation is spent discussing how to use that parameter in a group by context.

- Finally, I have absolutely no idea what contract `map_batches` imposes on its `function` arguments, and whether that contract depends on the context in which it is used.

**Comment by braaannigan** [comment]
Good points @soerenwolfers - I'd be happy to help if you want to make a PR

**Comment by soerenwolfers** [comment]
I would need to be certain about what's going first, and I think beyond playing with the functions to see what's currently possible I would also need some guidance on what the intentions are.

For example, I don't see the need yet to have more than 
- On the one hand, "elementwise mappings", the functions supplied to which would have to operate element-/row-wise but are allowed to set a `vectorized` parameter that determines whether the supplied function is `() -> ()` or `(n) -> (n)`. 
That's what I would call `map_rows`. 
On expressions, the inputs would either be individual values of a single series (e.g. `pl.col('a').map_rows(f, vectorized=False)`) or arbitrarily sized chunks of series (e.g. `pl.col('a').map_rows(f, vectorized=True)`). On dataframes the inputs would either be individual rows (e.g., `df.map_rows(f, vectorized=False)`) or arbitrarily sized chunks of a series of row-type (e.g., `df.map_rows(f, vectorized=True)`). 
- On the other hand, "groupwise mappings", the functions supplied to which would not need to satisfy any contract except needing to be `(n)->(n)` in a window context and `(n)->()` in a groupby context. That's what I would call `map_groups`. On expressions (e.g., `pol.col('a').map_groups(f).over('b')`) the inputs would be chunks (corresponding to the *entire* respective window partition) of a series. On dataframes, the inputs would be chunks of a dataframe (corresponding to the *entire* respective group) (e.g., `df.group_by('b').map_groups(f, [pl.col('a'), pl.col('b')])` (the second argument defaulting to `pl.all()`)) or sub-chunks (corresponding to the entire respective group) of a series (e.g., `df.group_by('b').agg(pl.col('a').map_groups(f))`)

**Comment by braaannigan** [comment]
I can't speak for the devs, but I don't think there's any drive to change the names or API, I think improving the documentation would be enough

**Comment by cmdlineluser** [comment]
There is also the top-level [`pl.map_groups`](https://docs.pola.rs/docs/python/dev/reference/expressions/api/polars.map_groups.html#polars-map-groups) (and `map_batches`, but no "elements")

`.map_elements` changes behaviour depending on the context i.e. it is "map_rows" in select context but inside a groupby context it is the same as `.map_groups` (without the outer `[]`)?

```python
import json
import polars as pl

df = pl.DataFrame(dict(
    group=[1, 2, 3, 1, 2], 
    value=[1, 2, 3, 4, 5]
))

df.with_columns(
    elements = pl.col("value").map_elements(lambda x: json.dumps(list(x))).over("group"),
    groups   = pl.map_groups("value", lambda x: json.dumps(list(x[0]))).over("group")
)
```

```python
# shape: (5, 4)
# ┌───────┬───────┬──────────┬────────┐
# │ group ┆ value ┆ elements ┆ groups │
# │ ---   ┆ ---   ┆ ---      ┆ ---    │
# │ i64   ┆ i64   ┆ str      ┆ str    │
# ╞═══════╪═══════╪══════════╪════════╡
# │ 1     ┆ 1     ┆ [1, 4]   ┆ [1, 4] │
# │ 2     ┆ 2     ┆ [2, 5]   ┆ [2, 5] │
# │ 3     ┆ 3     ┆ [3]      ┆ [3]    │
# │ 1     ┆ 4     ┆ [1, 4]   ┆ [1, 4] │
# │ 2     ┆ 5     ┆ [2, 5]   ┆ [2, 5] │
# └───────┴───────┴──────────┴────────┘
```

Should there not just be `Expr.map_rows` / `Expr.map_groups` and disallow `.map_rows` in a groupby context?

**Comment by vadimcn** [comment]
On top of this, there seem to be 3 flavors of `map_batches`, all slightly different:
-  [polars.map_batches](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.map_batches.html#polars.map_batches), 
- [Expr.map_batches](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.map_batches.html#polars.Expr.map_batches), 
- [LazyFrame.map_batches](https://docs.pola.rs/api/python/stable/reference/lazyframe/api/polars.LazyFrame.map_batches.html#polars.LazyFrame.map_batches).    

Why `LazyFrame.map_batches` takes a `DataFrame`, while others take a `Series`?   Nevertheless, it seems that the `DataFrame` that the function is called with, always contains only one column?

Why isn't there `streamable` of `is_elementwise` parameter for `polars.map_batches`?

According to documentation, `polars.map_batches` may be applied to multiple expr's at once, but will the function be called with each series separately, or all at once (as multiple arguments)?

Also, why isn't there `DataFrame.map_batches`?  

I would concur with the OP: `map_...` APIs are in a serious need of rationalization.

**Comment by gkorepanov** [comment]
+1, documentation issues are thoroughly described here. I’d be happy to contribute a PR if devs could clarify their intentions for these functions. Currently, I can’t confidently address the questions raised by @soerenwolfers without digging through the source code—this itself strongly suggests the documentation needs improvement.

**Comment by gregglind** [comment]
I also would love to be involved in documenting the `map` functions, including recipes for avoiding them, which include:

* expressions
* ufuncs
* `replace_strict`

Documenting these is a *big project*.  

Is it true that the ask here to the devs is:  "If people do this, can we expect there will be a little support in helping to understand what is supposed to be happeing in the api?"

**Comment by deanm0000** [comment]
ufuncs only avoids you having to type `map_batches`. It still dispatches `map_batches` in the background so it's only a convenience tool, not a performance one. Additionally, some numpy functions are faster than the polars version so `map_batches` isn't inherently slower and to always be avoided. For example:

```python
df=pl.DataFrame({"a":np.random.normal(0,5,10_000_000)})

%%timeit
df.select(pl.col("a").cum_sum())
## 89.5 ms ± 10.2 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
df.select(pl.col("a").map_batches(lambda x: np.cumsum(x.to_numpy())))
## 44 ms ± 7.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

One caveat is that with map_batches you've got GIL blocking so if you were doing that across a bunch of columns the polars expression version could do all the columns in parallel whereas the map_batches/numpy version has to do them sequentially.

On the other hand `map_elements` should always be avoided when possible because it is never vectorized, equivalent to a normal python for loop.

The difference between them is that map_batches is going to give the entire column to the function as a single input whereas map_elements is going to give every single value from the column to the function one at a time.

**Comment by deanm0000** [comment]
Those comments are mostly outdated.https://github.com/pola-rs/polars/issues/13557
